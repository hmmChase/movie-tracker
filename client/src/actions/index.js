export const addMovies = movies => ({
  type: 'ADD_MOVIES',
  movies
});

export const toggleLogin = () => ({
  type: 'TOGGLE_LOGIN'
});

export const storeUserData = ({ id, name }) => ({
  type: 'STORE_USER_DATA',
  id,
  name
});

export const loadFavorites = favorites => ({
  type: 'LOAD_FAVORITES',
  favorites
});