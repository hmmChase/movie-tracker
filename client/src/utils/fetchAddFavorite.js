export const fetchAddFavorite = async (movie, user_id) => {
  try {
    const url = '/api/users/favorites/new';
    const options = {
      method: 'POST',
      body: JSON.stringify({
        ...movie,
        user_id
      }),
      headers: { 'Content-Type': 'application/json' }
    };
    const response = await fetch(url, options);

    if (!response.ok) {
      throw new Error(`${response.status}`);
    }
    return response;
  } catch (error) {
    throw new Error(`Network request failed. (error: ${error.message})`);
  }
};
