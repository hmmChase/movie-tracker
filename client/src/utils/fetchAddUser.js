export const fetchAddUser = async user => {
  try {
    const url = '/api/users/new';
    const options = {
      method: 'POST',
      body: JSON.stringify(user),
      headers: { 'Content-Type': 'application/json' }
    };
    const response = await fetch(url, options);
    if (!response.ok) {
      throw new Error(`${response.status}`);
    }
    return response;
  } catch (error) {
    throw new Error(`Network request failed. (error: ${error.message})`);
  }
};
