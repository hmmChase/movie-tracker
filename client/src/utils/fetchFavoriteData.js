import { doFetch } from './doFetch';

export const fetchFavoriteData = async user_id => {
  const url = `/api/users/${user_id}/favorites`;
  const favoritesList = await doFetch(url);
  return favoritesList.data;
};
