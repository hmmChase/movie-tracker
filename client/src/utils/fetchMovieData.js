import { apiKey } from '../private/apiKey';
import { doFetch } from './doFetch';

export const fetchMovieData = async () => {
  const url = 'https://api.themoviedb.org/3/movie/now_playing?api_key=';
  const dirtyMovieData = await doFetch(url + apiKey);
  return cleanMovieData(dirtyMovieData.results);
};

export const cleanMovieData = movieData => {
  return movieData.map(movie => ({
    movie_id: movie.id,
    title: movie.title,
    poster_path: movie.poster_path,
    release_date: movie.release_date,
    vote_average: movie.vote_average,
    overview: movie.overview
  }));
};
