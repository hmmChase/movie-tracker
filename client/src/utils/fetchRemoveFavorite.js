export const fetchRemoveFavorite = async (user_id, movie_id) => {
  try {
    const url = `/api/users/${user_id}/favorites/${movie_id}`;
    const options = { method: 'DELETE' };
    const response = await fetch(url, options);

    if (!response.ok) {
      throw new Error(`${response.status}`);
    }
    return response;
  } catch (error) {
    throw new Error(`Network request failed. (error: ${error.message})`);
  }
};
