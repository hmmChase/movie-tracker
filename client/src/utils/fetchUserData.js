import { doFetch } from './doFetch';

export const fetchUserData = async () => await doFetch('/api/users');
