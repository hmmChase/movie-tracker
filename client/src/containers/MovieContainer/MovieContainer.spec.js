import React from 'react';
import { shallow } from 'enzyme';
import {
  MovieContainer,
  mapStateToProps,
  mapDispatchToProps
} from './MovieContainer';
import { addMovies, loadFavorites } from '../../actions';
import { fetchMovieData } from '../../utils/fetchMovieData';
jest.mock('../../utils/fetchMovieData');
import { fetchAddFavorite } from '../../utils/fetchAddFavorite';
jest.mock('../../utils/fetchAddFavorite');
import { fetchFavoriteData } from '../../utils/fetchFavoriteData';
jest.mock('../../utils/fetchFavoriteData');
import { fetchRemoveFavorite } from '../../utils/fetchRemoveFavorite';
jest.mock('../../utils/fetchRemoveFavorite');

describe('MovieContainer', () => {
  let movieContainer;
  let mockProps;

  beforeEach(() => {
    jest.resetAllMocks();
    mockProps = {
      movies: [
        {
          movie_id: 147,
          poster_path: '/movie-path/poster.png',
          title: 'mock title',
          release_date: '12-45-68',
          vote_average: 5,
          overview: 'mock overview'
        }
      ],
      favorites: [
        {
          movie_id: 421,
          poster_path: '/movie-path/poster.png',
          title: 'mock title',
          release_date: '12-45-68',
          vote_average: 5,
          overview: 'mock overview'
        }
      ],
      loggedIn: false,
      userId: 1,
      addMovies: jest.fn(),
      loadFavorites: jest.fn(),
      location: { pathname: '/' }
    };

    movieContainer = shallow(<MovieContainer {...mockProps} />, {
      disableLifecycleMethods: true
    });
  });

  it('matches the snapshot', () => {
    expect(movieContainer).toMatchSnapshot();
  });

  describe('componentDidMount', () => {
    it('calls getMovieData', () => {
      const mockGetMovieData = (movieContainer.instance().getMovieData = jest.fn());
      movieContainer.instance().componentDidMount();

      expect(mockGetMovieData).toHaveBeenCalledTimes(1);
    });
  });

  describe('getMovieData', () => {
    it('calls fetchMovieData', async () => {
      await movieContainer.instance().getMovieData();

      expect(fetchMovieData).toHaveBeenCalledTimes(1);
    });

    it('calls addMovies action', async () => {
      await movieContainer.instance().getMovieData();

      expect(mockProps.addMovies).toHaveBeenCalledTimes(1);
    });
  });

  describe('toggleFavorite', () => {
    it('if loggedIn is true, and checkIfFavorite is true, calls removeFavorite , ', () => {
      mockProps.loggedIn = true;
      movieContainer = shallow(<MovieContainer {...mockProps} />, {
        disableLifecycleMethods: true
      });
      movieContainer.instance().checkIfFavorite = jest
        .fn()
        .mockImplementation(() => true);
      const removeFavorite = (movieContainer.instance().removeFavorite = jest.fn());

      movieContainer.instance().toggleFavorite();

      expect(removeFavorite).toHaveBeenCalledTimes(1);
    });

    it('if loggedIn is true, and checkIfFavorite is false, calls addFavorite, ', () => {
      mockProps.loggedIn = true;
      movieContainer = shallow(<MovieContainer {...mockProps} />, {
        disableLifecycleMethods: true
      });
      const checkIfFavorite = (movieContainer.instance().checkIfFavorite = jest.fn());
      const addFavorite = (movieContainer.instance().addFavorite = jest.fn());
      movieContainer.instance().toggleFavorite();

      expect(checkIfFavorite).toHaveBeenCalledTimes(1);
      expect(addFavorite).toHaveBeenCalledTimes(1);
    });

    it('calls alert if user loggedIn is false', () => {
      window.alert = jest.fn();
      movieContainer.instance().toggleFavorite();

      expect(window.alert).toHaveBeenCalledTimes(1);
      expect(window.alert).toHaveBeenCalledWith(
        'You must be logged in to add favorites'
      );
    });
  });
  describe('checkIfFavorite', () => {
    it('returns false is if movie_id is not found in favorites', () => {
      const movie_id = 123;
      mockProps.favorites = [{ movie_id: 321 }];
      movieContainer = shallow(<MovieContainer {...mockProps} />, {
        disableLifecycleMethods: true
      });
      const checkIfFavorite = movieContainer
        .instance()
        .checkIfFavorite(movie_id);

      expect(checkIfFavorite).toBe(false);
    });

    it('returns true is if movie_id is found in favorites', () => {
      const movie_id = 123;
      mockProps.favorites = [{ movie_id: 123 }];
      movieContainer = shallow(<MovieContainer {...mockProps} />, {
        disableLifecycleMethods: true
      });
      const checkIfFavorite = movieContainer
        .instance()
        .checkIfFavorite(movie_id);

      expect(checkIfFavorite).toBe(true);
    });
  });

  describe('addFavorite', () => {
    it('calls fetchAddFavorite', async () => {
      await movieContainer.instance().addFavorite();

      expect(fetchAddFavorite).toHaveBeenCalledTimes(1);
    });

    it('calls fetchFavoriteData', async () => {
      await movieContainer.instance().addFavorite();

      expect(fetchFavoriteData).toHaveBeenCalledTimes(1);
    });

    it('calls loadFavorites', async () => {
      await movieContainer.instance().addFavorite();

      expect(mockProps.loadFavorites).toHaveBeenCalledTimes(1);
    });
  });

  describe('removeFavorite', () => {
    it('calls fetchRemoveFavorite', async () => {
      await movieContainer.instance().removeFavorite();

      expect(fetchRemoveFavorite).toHaveBeenCalledTimes(1);
    });

    it('calls fetchFavoriteData', async () => {
      await movieContainer.instance().removeFavorite();

      expect(fetchFavoriteData).toHaveBeenCalledTimes(1);
    });

    it('calls loadFavorites', async () => {
      await movieContainer.instance().removeFavorite();

      expect(mockProps.loadFavorites).toHaveBeenCalledTimes(1);
    });
  });
  describe('movieCards', () => {
    it('returns MovieCard components for each movie', () => {
      const movieCards = movieContainer.instance().movieCards();

      expect(movieCards).toHaveLength(1);
    });

    it('returns MovieCard components for each favorite', () => {
      mockProps.location = { pathname: '/favorites' };
      movieContainer = shallow(<MovieContainer {...mockProps} />, {
        disableLifecycleMethods: true
      });
      const movieCards = movieContainer.instance().movieCards();

      expect(movieCards).toHaveLength(1);
    });
  });

  describe('mapStateToProps', () => {
    it('maps state properties to props', () => {
      const mockState = {
        movies: [],
        user: {
          loggedIn: false,
          id: 1,
          favorites: []
        }
      };
      const expected = {
        movies: [],
        loggedIn: false,
        userId: 1,
        favorites: []
      };
      const mappedProps = mapStateToProps(mockState);

      expect(mappedProps).toEqual(expected);
    });
  });

  describe('mapDispatchToProps', () => {
    it('calls dispatch with the correct params on addMovies', () => {
      const mockDispatch = jest.fn();
      const mappedProps = mapDispatchToProps(mockDispatch);
      const movies = [];
      const mockAction = addMovies(movies);
      mappedProps.addMovies(movies);

      expect(mockDispatch).toHaveBeenCalledWith(mockAction);
    });

    it('calls dispatch with the correct params on loadFavorites', () => {
      const mockDispatch = jest.fn();
      const mappedProps = mapDispatchToProps(mockDispatch);
      const favoritesArray = [];
      const mockAction = loadFavorites(favoritesArray);
      mappedProps.loadFavorites(favoritesArray);

      expect(mockDispatch).toHaveBeenCalledWith(mockAction);
    });
  });
});
