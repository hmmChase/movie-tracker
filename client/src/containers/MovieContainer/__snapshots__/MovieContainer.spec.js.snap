// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieContainer matches the snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <MovieContainer
    addMovies={[Function]}
    favorites={
        Array [
            Object {
              "movie_id": 421,
              "overview": "mock overview",
              "poster_path": "/movie-path/poster.png",
              "release_date": "12-45-68",
              "title": "mock title",
              "vote_average": 5,
            },
          ]
    }
    loadFavorites={[Function]}
    location={
        Object {
            "pathname": "/",
          }
    }
    loggedIn={false}
    movies={
        Array [
            Object {
              "movie_id": 147,
              "overview": "mock overview",
              "poster_path": "/movie-path/poster.png",
              "release_date": "12-45-68",
              "title": "mock title",
              "vote_average": 5,
            },
          ]
    }
    userId={1}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <MovieCard
          isFavorite={[Function]}
          movie_id={147}
          overview="mock overview"
          poster_path="/movie-path/poster.png"
          release_date="12-45-68"
          title="mock title"
          toggleFavorite={[Function]}
          vote_average={5}
/>,
      ],
      "className": "MovieContainer",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": "147",
        "nodeType": "function",
        "props": Object {
          "isFavorite": [Function],
          "movie_id": 147,
          "overview": "mock overview",
          "poster_path": "/movie-path/poster.png",
          "release_date": "12-45-68",
          "title": "mock title",
          "toggleFavorite": [Function],
          "vote_average": 5,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "main",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <MovieCard
            isFavorite={[Function]}
            movie_id={147}
            overview="mock overview"
            poster_path="/movie-path/poster.png"
            release_date="12-45-68"
            title="mock title"
            toggleFavorite={[Function]}
            vote_average={5}
/>,
        ],
        "className": "MovieContainer",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": "147",
          "nodeType": "function",
          "props": Object {
            "isFavorite": [Function],
            "movie_id": 147,
            "overview": "mock overview",
            "poster_path": "/movie-path/poster.png",
            "release_date": "12-45-68",
            "title": "mock title",
            "toggleFavorite": [Function],
            "vote_average": 5,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "main",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
    "disableLifecycleMethods": true,
  },
}
`;
